### ISSUE TITLE
false positive  for numpy ufunc: bad-string-format-type

### ISSUE THREAD

**Issue by bpedersen2** [PyCQA/pylint#5533]
### Bug description

```
from numpy import radians, degrees #  any ufunc will do
  
a=radians(90.0)
print ( "%6.3f"%( degrees(a),))
print ( "%6.3f"% degrees(a))

# ==>  [E1307(bad-string-format-type)] Argument '.ndarray' does not match format type 'f'

# even though the ufunc returns a float-compat arg in this case

# Note: using f-strings this is accepted:

print ( f"{degrees(a):6.3f}" )
```

~                                       

### Command used

```shell
pylint check_pylint.py
```

### Pylint output

```shell
************* Module check_pylint
check_pylint.py:4: [E1307(bad-string-format-type), ] Argument '.ndarray' does not match format type 'f'
```
```


### Expected behavior

No warning as the ufunc result is usable in this case

Note: As can be seen, this warning is only displayed if passing a list to %-formatting.

### Pylint version

```shell
pylint 2.12.2
astroid 2.9.0
Python 3.8.10 (default, Nov 26 2021, 20:14:08) 
[GCC 9.3.0]
```

### OS / Environment

ubuntu 20.04

**Comment by orSolocate** [comment]
@bpedersen2  note that for the same line you also get this `[C]` message:

`C0209: Formatting a regular string which could be a f-string (consider-using-f-string)`

(same Pylint and astroid versions)

**Comment by orSolocate** [comment]
I debugged your session.
The `degrees` function returns an array `ndarray` or `None`, unless in the special case that you have here where the input is a scalar, which yields a scalar output as well.

From `numpy` `degrees` docs:

```
    x : array_like
        Input array in radians.
    Returns
    -------
    y : ndarray of floats
        The corresponding degree values; if `out` was supplied this is a
        reference to it.
        This is a scalar if `x` is a scalar.
```

running your code on Pylint, you get for **string.py**:

`#400:` `                   arg_type = utils.safe_infer(arg)`

assigns: `<Instance of .ndarray at 0x1613297147720>`

Conclusion: It is a Pylint bug caused by a astroid bug in `infer` function..


I can exclude `ndarray` arg_type from the `bad-string-format-type` it will solve this case, but might increase false positives regarding to this type

**Comment by mccaidan** [comment]
Why hasn't this change been implemented? I have a school project where I need to make pull requests to an open-source project, and I saw this was a good first issue. Could it be assigned to me to fix? If it doesn't need fixing, who can I ask for recommendations on what issues to fix?

**Comment by Pierre-Sassoulas** [comment]
Hey @mccaidan, it's just a lack of time I guess. Thank you for being willing to work on pylint. The label used by github to indicate good first issue is "good first issue" we also have "help wanted" in pylint. Be careful not all 'good first issue' are easy we sometime underestimate the difficulty. You can also sort by thumbs up reactions to work on the one that have more impact: https://github.com/PyCQA/pylint/issues?q=is%3Aopen+label%3A%22good+first+issue%22+sort%3Areactions-%2B1-desc

**Comment by mccaidan** [comment]
Is the solution not just the fix you mentioned in the comment?

**Comment by Pierre-Sassoulas** [comment]
@orSolocate did most of the diagnosis already, but there might be issue during implementation. You'll need to modify astroid, the library pylint is using. It's here : https://github.com/PyCQA/astroid

**Comment by mccaidan** [comment]
I now have time to work on this for the next couple of days. I just cloned
pylint and ran with pytest the tests that were included. Is it expected to
have 3 fail atm? Also, what would be a better way to communicate? The
discord?

**Comment by Pierre-Sassoulas** [comment]
>  Is it expected to have 3 fail atm?

No, the latest commit ran correctly https://github.com/PyCQA/pylint/commit/e087b4140615e2166b4705bfbbc226494eb90059 (except for a server error for the coverage). Did you install the dependencies from the requirements file ?

**Comment by mccaidan** [comment]
I have done
pip install -r [every .txt file with requirements in the title]
for both pylint and astroid in their most recent commit. Astroid passes all tests, but pylint still fails test_functional.py::test_functional[forgotten_debug_statement_py37] test_self.py::TestRunTC::test_do_not_import_files_from_local_directory_with_pythonpath
when using
python3 -m pytest tests/
When I run
python3 -m tox --recreate
it fails the pylint test at the beginning, but passes all the test_functional.py ones and fails 10 of the test_self.py::TestCallbackOptions ones.
Any tips on how to resolve this?

Also, if I make a change in the astroid directory code will it be reflected in pylint tests? If not, how do I integrate astroid changes into pylint when testing?

**Comment by jacobtylerwalls** [comment]
cd to the Astroid dir you're editing and:
pip install -e .

(Do this after doing the same for pylint bc doing it in the other order will cause pylint to give you the wrong astroid again)

**Comment by Pierre-Sassoulas** [comment]
It's also possible that the cutting edge astroid (2.12.0-dev0) do not pass all tests for pylint. Cutting edge pylint (2.14.0-dev0) is guaranteed to work with astroid 2.11.3. It should work, but we're not checking for each commit. Could you provide the output of pytest for the failed tests and your environment if the problem persists, please ?

**Comment by mccaidan** [comment]
I uninstalled python3.8/site-packages/astroid.egg-link with
pip uninstall astroid
and installed the 2.11.3 version with
pip install astroid==2.11.3
and ran the tests again, but they still fail. Is it because I didn't remove the astroid directory that I cloned?
![image](https://user-images.githubusercontent.com/59668503/164554951-e6de515b-17f4-4026-a010-19fe74232579.png)
printenv
SHELL=/bin/bash
NVM_INC=/home/aidan/.nvm/versions/node/v16.14.0/include/node
WSL_DISTRO_NAME=Ubuntu
NAME=DESKTOP-9KNGTBS
PWD=/home/aidan/src/githubopensource/pylint
LOGNAME=aidan
HOME=/home/aidan
LANG=C.UTF-8
WSL_INTEROP=/run/WSL/1396_interop

![image](https://user-images.githubusercontent.com/59668503/164556020-5119064f-6fa0-44cb-a0be-0bcf52bfcdc6.png)

![image](https://user-images.githubusercontent.com/59668503/164556061-7c6dcf3b-d646-4685-9244-02ffeb680e50.png)

**Comment by Pierre-Sassoulas** [comment]
It look like you have a ``no-member`` on the line ``pdb.set_trace()  # [forgotten-debug-statement]``, I don't know exactly what could be wrong, could you try to open an interpreter and do 
```python
import pdb
pdb.set_trace()
```
in it ?

**Comment by mccaidan** [comment]
![image](https://user-images.githubusercontent.com/59668503/164771710-086cc118-0de9-4761-8bd5-0c30e1e94b29.png)
Ok, it seemed to work, what now?

**Comment by orSolocate** [comment]
@mccaidan  - Did you install your forked repo on your machine according to the guidelines in [the docs](https://pylint.pycqa.org/en/latest/development_guide/testing.html?highlight=testing#test-your-code):

> Pylint is very well tested and has a high code coverage. New contributions are not accepted unless they include tests.
> 
> Before you start testing your code, you need to install your source-code package locally. To set up your environment for testing, open a terminal outside of your forked repository and run:
> 
> `pip install -e <forked_repo_dir_name>`
> 
> This ensures your testing environment is similar to Pylint's testing environment on GitHub.

I added this to the docs because I had some trouble in windows and this seems to fix it all.

Are you working on a WINDOWS machine?

btw - I would not say it is an easy fix.
What I recommend you to do is:

-> Add the test-code in this post to a functional test in pylint
-> Debug the test and break (breakpoint) at the place where the **infer** is used.
-> Step Inside into `astroid` functions to understand why the infer goes wrong.
Use the info in my comment, it will be useful.
-> make the change in astroid to solve it.
-> reinstall new astroid,
-> run your pylint test again, see it works.
-> run all test to see you didn't break anything.
-> PR

**Comment by orSolocate** [comment]
@Pierre-Sassoulas  I see no one took care of this issue, could I just make pylint ignore this false positive case with my suggestion above?

before:
`if arg_type == "builtins.int": return True`

after:
`if arg_type in {"builtins.int", ".ndarray"}: return True`

**Comment by Pierre-Sassoulas** [comment]
I think this is something we should not fix in pylint but we should either make astroid understand the numpy's underlying code correctly, use a numpy brain (but we have a lot of those and they are buggy). We have a lot of numpy brain, but this is a poor substitution to astroid understanding the code or a dedicated plugin that will be able to handle multiple numpy version. The best solution is probably to defer to a pylint-numpy plugin that is in need of creation and probably also potential contributor. Deciding to close this kind of issue as "should be handled in an external plugin" would affect a LOT of issues (all the lib specific one) so this need to be discussed among contributors first. brain is the easy/fast fix but then we need to upgrade the brain when a new version of numpy is out and the code change and we don't have a lot of tests for numpy and none for multiple numpy version.
